tiled/server/router.py

```diff
@@ -1215,3 +1219,104 @@ async def delete_revision(
 
     await entry.delete_revision(number)
     return json_or_msgpack(request, None)
+
+
+@router.get("/asset/bytes/{path:path}")
+async def get_asset(
+    request: Request,
+    id: int,
+    relative_path: Optional[Path] = None,
+    entry=SecureEntry(scopes=["read:data"]),  # TODO: Separate scope for assets?
+):
+    if not hasattr(entry, "asset_by_id"):
+        raise HTTPException(
+            status_code=405,
+            detail="This node does not support downloading assets.",
+        )
+
+    asset = await entry.asset_by_id(id)
+    if asset is None:
+        raise HTTPException(
+            status_code=404,
+            detail=f"This node exists but it does not have an Asset with id {id}",
+        )
+    if asset.is_directory:
+        if relative_path is None:
+            raise HTTPException(
+                status_code=400,
+                detail=(
+                    "This asset is a directory. Must specify relative path, "
+                    f"from manifest provided by /asset/manifest/...?id={id}"
+                ),
+            )
+        if relative_path.is_absolute():
+            raise HTTPException(
+                status_code=400,
+                detail="relative_path query parameter must be a *relative* path",
+            )
+    else:
+        if relative_path is not None:
+            raise HTTPException(
+                status_code=400,
+                detail="This asset is not a directory. The relative_path query parameter must not be set.",
+            )
+    if not asset.data_uri.startswith("file:"):
+        raise HTTPException(
+            status_code=400,
+            detail="Only download assets stored as file:// is currently supported.",
+        )
+    path = path_from_uri(asset.data_uri)
+    if relative_path is not None:
+        # Be doubly sure that this is under the Asset's data_uri directory
+        # and not sneakily escaping it.
+        if not os.path.commonpath([path, path / relative_path]) != path:
+            # This should not be possible.
+            raise RuntimeError(
+                f"Refusing to serve {path / relative_path} because it is outside "
+                "of the Asset's directory"
+            )
+        full_path = path / relative_path
+    else:
+        full_path = path
+    stat_result = await anyio.to_thread.run_sync(os.stat, full_path)
+    filename = full_path.name
+    return FileResponse(
+        full_path,
+        stat_result=stat_result,
+        method="GET",
+        status_code=200,
+        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
+    )
+
+
+@router.get("/asset/manifest/{path:path}")
+async def get_asset_manifest(
+    request: Request,
+    id: int,
+    entry=SecureEntry(scopes=["read:data"]),  # TODO: Separate scope for assets?
```

Comment:
Is it likely that the assets contain juicy information that is not already returned by read()?

Reply:
I can imagine there might be some use cases for taking the hard-line position that files are internal and not public API, so that file formats can be changed with 100% certainty that no user workflow could be broken.

I think we can wait to carve out finer scopes until this use case emerges. We might get there someday, or we might not.

Reply:
That makes sense. The new `expose_raw_assets` config parameter helps in this regard.